### PR TITLE
Add 'direction' option (prop) to 'withDirection' HOC, This to allow the overwrite of the direction when required

### PR DESCRIPTION
**Pull request by luis-mgb**
Add 'direction' option (prop) to 'withDirection' HOC, This to allow the overwrite of the direction when required (for phone numbers for example that are always LTR), updated packages

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,22 +1,32 @@
 # Change Log
 
 ## v1.3.0
+
+- [New] Add `direction` option (prop) to `withDirection` HOC, This to allow the overwrite of the direction when required (for phone numbers for example that are always LTR)
```
Comment (ljharb) [inline CHANGELOG.md]: v1.3.0 is already released; if you're going to add changelog items, please add in an "unreleased" section

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,22 +1,32 @@
...
 - [Dev Deps] update `babel-jest`, `chai`, `eslint`, `eslint-config-airbnb`, `eslint-plugin-jsx-a11y`, `eslint-plugin-react`, `jest`, `rimraf`, `sinon`, `sinon-chai`, `webpack`
 
 ## v1.0.1
- - [Fix] allow down to React v15.0
- - [Deps] update `airbnb-prop-types`, `deepmerge`, `hoist-non-react-statics`
- - [Dev Deps] update `babel-cli`, `casual`, `eslint`, `eslint-plugin-jsx-a11y`, `eslint-plugin-react`, `sinon-chai`, `sinon`, `webpack`; remove `eslint-config-airbnb-base`
- - [Dev Deps] add `babel-jest`, which needs to be explicitly a top-level dev dep for npm 2
- - [Tests] test on node 4, but not node 7
+
+- [Fix] allow down to React v15.0
```
Comment (ljharb) [inline CHANGELOG.md]: the space is intended; the bug is that later sections *lack* a space.

File: package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "react-with-direction",
-  "version": "1.3.0",
+  "version": "1.4.0",
```
Comment (ljharb) [inline package.json]: version numbers must never be bumped in PRs; separately, the deps should be a separate PR from this one. please revert these changes, especially the jest and babel ones.

File: src/withDirection.jsx [hunk 4]
```diff
@@ -30,9 +30,11 @@ export default function withDirection(WrappedComponent) {
   class WithDirection extends React.Component {
     constructor(props, context) {
       super(props, context);
-      this.state = {
-        direction: context[CHANNEL] ? context[CHANNEL].getState() : defaultDirection,
-      };
+      let direction = context[CHANNEL] ? context[CHANNEL].getState() : defaultDirection;
+      if (props.direction) {
```
Comment (ljharb) [inline src/withDirection.jsx:35]: if a prop is going to be added, a `.propTypes` entry *must* be added for it.

File: tests/AutoDirectionProvider_test.jsx [hunk 5]
```diff
@@ -9,43 +9,44 @@ import mockBrcast from './mocks/brcast_mock';
 
 describe('<AutoDirectionProvider>', () => {
   it('renders a DirectionProvider', () => {
-    const wrapper = shallow((
+    const wrapper = shallow(
       <AutoDirectionProvider text="a">
         <div />
-      </AutoDirectionProvider>
-    )).dive();
+      </AutoDirectionProvider>,
```
Comment (ljharb) [inline tests/AutoDirectionProvider_test.jsx]: please double-wrap in parens instead of using the commas - ie, please revert whatever autoformatting was applied.

**Review by ljharb — CHANGES_REQUESTED** [review]
I'm confused about the need for this prop. If you want to force the direction, wouldn't you wrap in a new `DirectionProvider`, with the direction set explicitly? Why is the prop needed?

File: CHANGELOG.md (hunk 1 above)
Comment (luis-mgb) [reply CHANGELOG.md]: Reverted the changes to the file, I'll update this once the PR is approved and bump the version on this package.

File: CHANGELOG.md (hunk 2 above)
Comment (luis-mgb) [reply CHANGELOG.md]: I reverted the changes

File: src/withDirection.jsx (hunk 4 above)
Comment (luis-mgb) [reply src/withDirection.jsx:35]: I updated the way props are assigned at the end of this file to address this, now direction is always an optional prop for the HOC

File: tests/AutoDirectionProvider_test.jsx (hunk 5 above)
Comment (luis-mgb) [reply tests/AutoDirectionProvider_test.jsx]: Reverted auto formatting on all files.

File: package.json (hunk 3 above)
Comment (luis-mgb) [reply package.json]: Reverted this changes

File: CHANGELOG.md (hunk 1 above)
Comment (ljharb) [reply CHANGELOG.md]: note; when merged, in a commit directly made to master.

File: src/withDirection.jsx [hunk 6]
```diff
@@ -23,16 +23,19 @@ const defaultDirection = DIRECTIONS.LTR;
 
 // export for convenience, in order for components to spread these onto their propTypes
 export const withDirectionPropTypes = {
-  direction: directionPropType.isRequired,
+  direction: directionPropType,
```
Comment (ljharb) [inline src/withDirection.jsx:26]: we went out of our way to leave this required when considering AutoDirectionProvider - the likelihood of a bug with someone forgetting to provide it is *very* high.

Anything wrapped in `withDirection` *must* get a direction, that's the sole purpose of the HOC.

File: src/withDirection.jsx [hunk 7]
```diff
@@ -23,16 +23,19 @@ const defaultDirection = DIRECTIONS.LTR;
...
 
 export default function withDirection(WrappedComponent) {
   class WithDirection extends React.Component {
     constructor(props, context) {
       super(props, context);
-      this.state = {
-        direction: context[CHANNEL] ? context[CHANNEL].getState() : defaultDirection,
-      };
+      let direction = context[CHANNEL] ? context[CHANNEL].getState() : defaultDirection;
+      // console.log(props.direction);
```
Comment (ljharb) [inline src/withDirection.jsx:34]:
```suggestion
```

File: src/withDirection.jsx [hunk 8]
```diff
@@ -68,8 +71,7 @@ export default function withDirection(WrappedComponent) {
   WithDirection.contextTypes = contextTypes;
   WithDirection.displayName = `withDirection(${wrappedComponentName})`;
   if (WrappedComponent.propTypes) {
-    WithDirection.propTypes = deepmerge({}, WrappedComponent.propTypes);
-    delete WithDirection.propTypes.direction;
+    WithDirection.propTypes = deepmerge(WrappedComponent.propTypes, withDirectionPropTypes);
```
Comment (ljharb) [inline src/withDirection.jsx:74]: it might be optional here, but it should stay required on the wrapped component.

**Review by ljharb — CHANGES_REQUESTED** [review]
Still curious about the above question:
> If you want to force the direction, wouldn't you wrap in a new DirectionProvider, with the direction set explicitly? Why is the prop needed?

**Comment by luis-mgb** [comment]
The issue was solved up-stream and this is no longer needed
